src/python/grpcio_tests/tests/unit/_compression_test.py

```diff
@@ -52,75 +113,248 @@ def __init__(self, request_streaming, response_streaming):
+        self._pre_response_callback = pre_response_callback
+
     def service(self, handler_call_details):
         if handler_call_details.method == _UNARY_UNARY:
-            return _MethodHandler(False, False)
+            return _MethodHandler(False, False, self._pre_response_callback)
+        elif handler_call_details.method == _UNARY_STREAM:
+            return _MethodHandler(False, True, self._pre_response_callback)
+        elif handler_call_details.method == _STREAM_UNARY:
+            return _MethodHandler(True, False, self._pre_response_callback)
         elif handler_call_details.method == _STREAM_STREAM:
-            return _MethodHandler(True, True)
+            return _MethodHandler(True, True, self._pre_response_callback)
         else:
             return None
 
 
+@contextlib.contextmanager
+def _instrumented_client_server_pair(channel_kwargs, server_kwargs,
+                                     server_handler):
+    host = '127.0.0.1'
+    server = grpc.server(futures.ThreadPoolExecutor(), **server_kwargs)
+    server.add_generic_rpc_handlers((server_handler,))
+    server_port = server.add_insecure_port('{}:0'.format(host))
+    server.start()
+    with _tcp_proxy.TcpProxy(host, host, server_port) as proxy:
+        proxy_port = proxy.get_port()
+        with grpc.insecure_channel('{}:{}'.format(host, proxy_port),
+                                   **channel_kwargs) as client_channel:
+            try:
+                yield client_channel, proxy, server
+            finally:
+                server.stop(None)
+
+
+def _get_byte_counts(channel_kwargs, multicallable_kwargs, client_function,
+                     server_kwargs, server_handler, message):
+    with _instrumented_client_server_pair(channel_kwargs, server_kwargs,
+                                          server_handler) as pipeline:
+        client_channel, proxy, server = pipeline
+        client_function(client_channel, multicallable_kwargs, message)
+        return proxy.get_byte_count()
+
+
+def _get_compression_ratios(client_function, first_channel_kwargs,
+                            first_multicallable_kwargs, first_server_kwargs,
+                            first_server_handler, second_channel_kwargs,
+                            second_multicallable_kwargs, second_server_kwargs,
+                            second_server_handler, message):
+    first_bytes_sent, first_bytes_received = _get_byte_counts(
+        first_channel_kwargs, first_multicallable_kwargs, client_function,
+        first_server_kwargs, first_server_handler, message)
+    second_bytes_sent, second_bytes_received = _get_byte_counts(
+        second_channel_kwargs, second_multicallable_kwargs, client_function,
+        second_server_kwargs, second_server_handler, message)
+    return ((second_bytes_sent - first_bytes_sent) / float(first_bytes_sent),
+            (second_bytes_received - first_bytes_received) /
+            float(first_bytes_received))
+
+
+_REQUEST = b'\x00' * 100
```

Comment:
Let's make constant definitions go to the top of the file? Apply to constants below.